### PR TITLE
Add changelog for 1.12.1-rc1

### PR DESCRIPTION
**Pull request by kunaltyagi**
[Step 2 from here](https://github.com/PointCloudLibrary/pcl/wiki/Preparing-Releases)

### CODE REVIEW

**Comment by kunaltyagi** [comment]
No mention of the ABI break. Should we change that?

File: CHANGES.md [hunk 1]
```diff
@@ -1,5 +1,103 @@
 # ChangeList
 
+## = 1.12.1 (2021.12.09) =
```
Comment (mvieth) [inline CHANGES.md:3]: Wouldn't this be Dec 16 or Dec 17 (pre-release + a week or so)? Or are you planning to change this when doing the actual release?

**Comment by kunaltyagi** [comment]
Release pipeline: https://dev.azure.com/PointCloudLibrary/pcl/_build/results?buildId=20517&view=results

File: CHANGES.md [hunk 2]
```diff
@@ -1,5 +1,103 @@
...
+
+#### libpcl_filters:
+
+* Clear the output indices in function `CropHull::applyFilters` [[#4851](https://github.com/PointCloudLibrary/pcl/pull/4851)]
+* Fix unresolved linking to Convolution [[#4845](https://github.com/PointCloudLibrary/pcl/pull/4845)]
+* **[deprecation]** Add deprecation for filter getters with bool reference [[#4861](https://github.com/PointCloudLibrary/pcl/pull/4861)]
+* NormalSpaceSampling filter: add constructor to specify `extract_removed_indices`  [[#4846](https://github.com/PointCloudLibrary/pcl/pull/4846)]
+* **[deprecation]** Fix keep_organized behavior in CropHull filter [[#4855](https://github.com/PointCloudLibrary/pcl/pull/4855)]
+* Added reserve function before storing points in PointCloud in VoxelGr… [[#4938](https://github.com/PointCloudLibrary/pcl/pull/4938)]
+* Add `PCL_MAKE_ALIGNED_OPERATOR_NEW` to CropBox for better Eigen support [[#4956](https://github.com/PointCloudLibrary/pcl/pull/4956)]
```
Comment (mvieth) [inline CHANGES.md:59]: This was reverted again, right?

**Review by mvieth — COMMENTED** [review]
Yes, mentioning the (possible) ABI break would be good

**Comment by mvieth** [comment]
And minor side note: it would be called 1.12.1-rc1, not 1.12.1-alpha, right?

File: CHANGES.md (hunk 2 above)
Comment (kunaltyagi) [reply CHANGES.md:59]: Yes, but we didn't remove the labels. Removed them now. Will generate the changelog again

**Review by larshg — DISMISSED** [review]
Looks good to me after @mvieth changes is applied.

Seems pcl_perception failed to build due to qhullstatic_r not being compiled wit -fPIC?

**Comment by kunaltyagi** [comment]
We shouldn't link to qhull static in that case

**Comment by mvieth** [comment]
I think I found the problem:
1. `PCL_QHULL_REQUIRED_TYPE` should be saved for downstream projects, just like [QHULL_USE_STATIC is](https://github.com/PointCloudLibrary/pcl/blob/master/PCLConfig.cmake.in#L132) (same for flann)
2. I think it is strange that the type defaults to STATIC ([here](https://github.com/PointCloudLibrary/pcl/blob/master/cmake/Modules/FindQhull.cmake#L140) and [here](https://github.com/PointCloudLibrary/pcl/blob/master/cmake/Modules/FindQhull.cmake#L69), same for flann) if PCL_QHULL_REQUIRED_TYPE matches neither DONTCARE nor SHARED. I think it should check SHARED and STATIC first, then treat everything else as DONTCARE

Let me know what you think, I can create a PR tomorrow

**Comment by larshg** [comment]
> I think I found the problem:
> 
> 1. `PCL_QHULL_REQUIRED_TYPE` should be saved for downstream projects, just like [QHULL_USE_STATIC is](https://github.com/PointCloudLibrary/pcl/blob/master/PCLConfig.cmake.in#L132) (same for flann)
> 2. I think it is strange that the type defaults to STATIC ([here](https://github.com/PointCloudLibrary/pcl/blob/master/cmake/Modules/FindQhull.cmake#L140) and [here](https://github.com/PointCloudLibrary/pcl/blob/master/cmake/Modules/FindQhull.cmake#L69), same for flann) if PCL_QHULL_REQUIRED_TYPE matches neither DONTCARE nor SHARED. I think it should check SHARED and STATIC first, then treat everything else as DONTCARE
> 
> Let me know what you think, I can create a PR tomorrow

1.
I don't actually see why we need to search for dependencies, that are required for building PCL when used downstream. If PCLs dependencies are statically linked into PCL, it will be present there.
If its lined dynamic, we don't need to find dependencies again to link against downstream projects. Only PCL libraries are needed. However, when you launch your application and a .so/.dll is not found and are required, ie. Qhull it will fail to load the application and tell that it could find qhull_r.dll (at least on windows).

Since All-In-One installer has all dlls dependencies this should work fine. PCL Linux package only contain PCLs .so, but has dependency to other packages, ie. Qhull which in turn install .so files and add them to some bin folder, where programs can locate them on launch?

2.
The logic should be like this:
If both static and shared is found, we use PCL_SHARED_LIBRARY to decide if REQUESTED_LIBRARY is DONTCARE.
Else we select shared if the requested is SHARED.
Else we select static (last option)

If only one of them is found, ie. static we select that one. If only shared is found, we select that one, no matter what Requested Library is set to. But when we return to main CMakeList it check if has found one that matches Requsted Library.

3.
It doesn't have anything to do with that Qhull is compiled without the -fPIC flag. I have opened a issue at Qhull: https://github.com/qhull/qhull/issues/107

4. What docker is used for build pcl_perceptron? Is both static and shared installed when installing libqhull-dev? It should default to shared version, if we are building PCL as shared.
Where can we supply and request Qhull as shared specifically - I couldn't quite figure it out in the docker build files - should it be set as argument to the cmake call?

**Comment by mvieth** [comment]
1. After the release, we could discuss again whether it is necessary to search for PCL's dependencies in downstream projects, but for now, I would prefer a more conservative solution with minimal chance of breaking things (adding `set(PCL_QHULL_REQUIRED_TYPE @PCL_QHULL_REQUIRED_TYPE@)` to PCLConfig.cmake.in)
2. I mean the case when qhull is available as shared _and_ as static, then there is basically a switch-statement on `PCL_QHULL_REQUIRED_TYPE`, with cases DONTCARE and SHARED, and the default case (when neither DONTCARE nor SHARED matches) leads to static being used. I would restructure it so that there are cases for SHARED and STATIC, and everything else (default case) would be treated as DONTCARE
3. Good that you opened the issue, let's see what Qhull's maintainers think (but we still have to solve this ourselves for the upcoming release)
4. The dockerfile is [this one](https://github.com/PointCloudLibrary/pcl/blob/master/.dev/docker/perception_pcl_ros/Dockerfile), it is more or less based on Ubuntu 18.04. Yes, both shared and static qhull libraries are available. The problem is that `PCL_QHULL_REQUIRED_TYPE` is empty in the downstream project (pcl_ros) because we don't set it, so it matches neither `DONTCARE` nor `SHARED`, so it defaults to the static library

**Comment by kunaltyagi** [comment]
I think we should use the logic that @larshg  pointed out:
1. If both found, with a preference for the library, use the preferred version
2. If both found, and no preference for library, use `PCL_LIB_LIBS` (which is `SHARED` by default, set via `PCL_SHARED_LIBS`)
3. If found just 1, select that
4. Check if preference for library is present, if so, check if it is same as the selected type, else error

> is it necessary to search for PCL's dependencies in downstream projects?
Usually, I don't see a reason to search for dependencies again, but I could be wrong

---
1. Would it be possible to switch to this logic for libraries before the release? Or should we change the logic for just QHull for now?
2. QHull github seems a bit slow, so I'd recommend going for a pcl-cmake solution rather than waiting on the upstream

**Comment by mvieth** [comment]
Actually my idea for the logic isn't that different, maybe I just described it badly. See my new PR for my suggestion.

> Would it be possible to switch to this logic for libraries before the release? Or should we change the logic for just QHull for now?

I strongly believe that we should keep the changes small/conservative for this minor release, and delay possibly more risky changes until later (with more time for testing)

> QHull github seems a bit slow, so I'd recommend going for a pcl-cmake solution rather than waiting on the upstream

True

**Comment by kunaltyagi** [comment]
Try `#2`: https://dev.azure.com/PointCloudLibrary/pcl/_build/results?buildId=20531&view=results

**Comment by kunaltyagi** [comment]
Pipeline 🟢

**Comment by kunaltyagi** [comment]
[Ready for rc1 release post merge](https://github.com/PointCloudLibrary/pcl/releases/edit/untagged-1abc935332e884d848f8)

**Review by larshg — APPROVED** [review]

**Review by mvieth — APPROVED** [review]

**Pull request merged**